modules/system/classes/FileManifest.php

```diff
@@ -0,0 +1,171 @@
+ * @author Ben Thomson
+ */
+class FileManifest
+{
+    /**
+     * @var string Root folder of this installation.
+     */
+    protected $root;
+
+    /**
+     * @var string Modules to store in manifest.
+     */
+    protected $modules = ['system', 'backend', 'cms'];
+
+    /**
+     * @var array Files cache.
+     */
+    protected $files = [];
+
+    /**
+     * Constructor.
+     *
+     * @param string $root The root folder to get the file list from.
+     * @param array $modules An array of modules to include in the file manifest.
+     */
+    public function __construct($root = null, array $modules = null)
+    {
+        if (isset($root)) {
+            $this->setRoot($root);
+        } else {
+            $this->setRoot(base_path());
+        }
+
+        if (isset($modules)) {
+            $this->setModules($modules);
+        } else {
+            $this->setModules(Config::get('cms.loadModules', ['System', 'Backend', 'Cms']));
+        }
+    }
+
+    /**
+     * Sets the root folder.
+     *
+     * @param string $root
+     * @throws ApplicationException If the specified root does not exist.
+     */
+    public function setRoot($root)
+    {
+        if (is_string($root)) {
+            $this->root = realpath($root);
+
+            if ($this->root === false || !is_dir($this->root)) {
+                throw new ApplicationException(
+                    'Invalid root specified for the file manifest.'
+                );
+            }
+        }
+
+        return $this;
+    }
+
+    /**
+     * Sets the modules.
+     *
+     * @param array $modules
+     */
+    public function setModules(array $modules)
+    {
+        $this->modules = array_map(function ($module) {
+            return strtolower($module);
+        }, $modules);
+
+        return $this;
+    }
+
+    /**
+     * Gets a list of files and their corresponding hashsums.
+     *
+     * @return array
+     */
+    public function getFiles()
+    {
+        if (count($this->files)) {
+            return $this->files;
+        }
+
+        $files = [];
+
+        foreach ($this->modules as $module) {
+            $path = $this->root . DIRECTORY_SEPARATOR . 'modules' . DIRECTORY_SEPARATOR . $module;
+
+            if (!is_dir($path)) {
+                continue;
+            }
+
+            foreach ($this->findFiles($path) as $file) {
+                $files[$this->getFilename($file)] = md5_file($file);
+            }
+        }
+
+        return $this->files = $files;
+    }
+
+    /**
+     * Gets the checksum of a specific install.
+     *
+     * @return array
+     */
+    public function getModuleChecksums()
+    {
+        if (!count($this->files)) {
+            $this->getFiles();
+        }
+
+        $modules = [];
+        foreach ($this->modules as $module) {
+            $modules[$module] = '';
+        }
+
+        foreach ($this->files as $path => $hash) {
+            // Determine module
+            $module = explode(DIRECTORY_SEPARATOR, $path)[2];
```

Comment:
DIRECTORY_SEPERATOR might be required here, but if you are able to test on a windows machine for sure, that would be helpful.